source/dotnet/Library/AdaptiveCards.Templating/AdaptiveCardsTemplateVisitor.cs

```diff
@@ -241,8 +228,17 @@ public override AdaptiveCardsTemplateResult VisitTemplateData([NotNull] Adaptive
                 }
             }
             else
+            // else clause handles all of the ordinary json values 
             {
-                throw new Exception($"'{context.value().GetText()}' at line, '{context.Start.Line}' is invalid value type for '$data : ' pair");
+                string childJson = templateDataValueNode.GetText();
+                try
+                {
+                    PushDataContext(childJson, root);
+                }
+                catch (Exception e)
+                {
+                    throw new Exception($"parsing data failed at line, '{context.Start.Line}', '{childJson}' was given", e);
```

Comment:
@jwoo-msft / @paulcam206,  should we be handling and re-throwing all System.Exception's or try to be specific about templating exceptions? What are the exceptions types that can result in the call to PushDataContext() ? Ideally we'd only be handling those.. 

Something like: 
try
{
    PushDataContext(childJson, root);
}
catch(Exception innerException)
{
 if (e is JsonSerializationException || e is InvalidOperationException) ==> or something similar
{
  throw new AdaptiveSerializationException("your string", innerException );
}
 throw; // rethrow the original "exceptional"/unexpected exception as-is to avoid overwriting original context/stacktrace.
}

Reply:
Btw we should add unit tests that invoke the exception code path also to test the non-golden-path behavior the customer will see